squidpy/gr/_ripley.py

```diff
@@ -0,0 +1,245 @@
+            distances_i, _ = tree_i.kneighbors(random, n_neighbors=1)
+            _, stats_i = _f_g_function(distances_i.squeeze(), support)
+        elif mode == "G":
+            tree_i = NearestNeighbors(metric=metric, n_neighbors=n_neigh).fit(random_i)
+            distances_i, _ = tree_i.kneighbors(coordinates, n_neighbors=1)
+            _, stats_i = _f_g_function(distances_i.squeeze(), support)
+        elif mode == "L":
+            distances_i = pdist(random_i, metric=metric)
+            _, stats_i = _l_function(distances_i, support, N, area)
+
+        for j in range(obs_arr.shape[0]):
+            pvalues[j] += stats_i >= obs_arr[j]
+        sims[i] = stats_i
+
+    pvalues /= n_simulations + 1
+    pvalues = np.minimum(pvalues, 1 - pvalues)
+
+    obs_df = _reshape_res(obs_arr.T, columns=le.classes_, index=bins, var_name=cluster_key)
+    sims_df = _reshape_res(sims.T, columns=np.arange(n_simulations), index=bins, var_name="simulations")
+
+    res = {f"{mode}_stat": obs_df, "sims_stat": sims_df, "bins": bins, "pvalues": pvalues}
+
+    if TYPE_CHECKING:
+        assert isinstance(res, dict)
+
+    if copy:
+        logg.info("Finish", time=start)
+        return res
+
+    _save_data(adata, attr="uns", key=Key.uns.ripley(cluster_key, mode), data=res)
+
+
+def _reshape_res(
+    results: np.ndarray, columns: Union[np.ndarray, List[str]], index: np.ndarray, var_name: str
+) -> pd.DataFrame:
+    df = pd.DataFrame(results, columns=columns, index=index)
+    df.index.set_names(["bins"], inplace=True)
+    df = df.melt(var_name=var_name, value_name="stats", ignore_index=False)
+    df[var_name] = df[var_name].astype("category", copy=True)
+    df.reset_index(inplace=True)
+    return df
+
+
+def _f_g_function(distances: np.ndarray, support: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
+    counts, bins = np.histogram(distances, bins=support)
+    fracs = np.cumsum(counts) / counts.sum()
+    return bins, np.concatenate((np.zeros((1,), dtype=np.float_), fracs))
+
+
+def _l_function(
+    distances: np.ndarray, support: np.ndarray, n: np.int_, area: np.float_
+) -> Tuple[np.ndarray, np.ndarray]:
+    n_pairs_less_than_d = (distances < support.reshape(-1, 1)).sum(axis=1)
+    intensity = n / area
+    k_estimate = ((n_pairs_less_than_d * 2) / n) / intensity
+    l_estimate = np.sqrt(k_estimate / np.pi)
+    return support, l_estimate
+
+
+def _ppp(hull: ConvexHull, n_simulations: int, n_observations: int, seed: Optional[int] = None) -> np.ndarray:
+    """
+    Simulate Poisson Point Process on a polygon.
+
+    Parameters
+    ----------
+    hull
+        Convex hull of the area of interest.
+    n_simulations
+        Number of simulated point processes.
+    n_observations
+        Number of observations to sample from each simulation
+    seed
+        Random seed.
+
+    Returns
+    -------
+        An Array with shape (n_simulation, n_observations, 2).
```

Comment:
Unnecessary indentation + please wrap the shape in double backticks as \`\`(n_simulation, n_observations, 2)\`\`.